### DISCUSSION TITLE
User Data on objects

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#79] [Q&A] [answered]
I would like to give a user an ability to associate some user defined number with an object. I can use user data on the body interface for myself (thank you for exposing it by the way!), and then save his number in a shape interface. However, that won't do for character virtual, as there is no body that I could use for internal mapping. It only has a shape. Is there another way I could work around it? I'd like to have one number for the framework's internal purposes, like mapping an object with the entity, and one for the user. I suppose I could pack two 16 bit numbers in a single 32 bit one, but wanted to check, if there is another option.

**Comment by LeXXik** [comment]
Thinking about it, I will change the indexing method to the actual body IDs, then I won't need user data and can leave it for users. I just need to think a bit how to do the mapping on the frontend framework, as it doesn't know about those IDs when the components are created.

**Comment by jrouwe** [comment]
You can use the user data on bodies / constraints / shapes in any way you like, Jolt is not using them.

The reason why CharacterVirtual doesn't have a user data is that it is not part of the simulation, so the only code that is aware of its existence / can receive callback is the code that interacts with a specific CharacterVirtual. So the caller should already be aware of the context. (And CharacterVirtual also doesn't have a BodyID)

**Comment by LeXXik** [comment]
Yeah, that's the idea. However, the context might not always be available. For example, I'm trying to implement your character controller example. You are using the `allowSliding` variable inside the `OnContactSolve` closure:

https://github.com/jrouwe/JoltPhysics.js/blob/078a0e42541f39704eb5b74f06177a17ad78d5d0/Examples/character_virtual.html#L147

This assumes that the closures are executed in the same context as the user script, so `allowSliding` is accessible. However, if the closures are sent to the webworker, I no longer have access to that variable (the scope gets lost). One option, is then to save a number, represeting truthy `allowSliding`, in a user data of the character's shape. Then the callback closure can read it from the character directly and decide on its logic.

**Reply by jrouwe** [reply, marked as answer]
Does `Jolt.getPointer(character)` give you a unique number that identifies which character it is? Or does `character` also live in different locations for the web worker / user script (sorry I never worked with web workers).

**Reply by LeXXik** [reply]
Right, in my setup Jolt can be optionally run in a web worker. In that case user scripts do not have a direct access to the physics systems and communicate with it via an interface. Inside the web worker the physics system has access to Jolt instance and characters. Main thread can access only user scripts. You can consider it as user scripts are running in one window, and the web worker code in another, and those windows can send each other messages, but cannot access each other's contexts.

I just tried `Jolt.getPointer(character)` with 2 chars and both give me unique IDs. I think I can leverage that. Thank you for the pointer! (pun intended)

**Reply by LeXXik** [reply]
This worked great. In the end, I changed the mapping logic a bit. I no longer store the entity-to-body index in user data, but instead keep a local map of front-end-index-to-jolt-pointer map. This allowed to free the user data for user scripts, and now my character no longer slides on slopes :)
